core/app/models/spree/promotion/rules/user_role.rb

```diff
@@ -0,0 +1,26 @@
+module Spree
+  class Promotion
+    module Rules
+      class UserRole < PromotionRule
+        has_many :promotion_rule_roles,
+          class_name: 'Spree::PromotionRuleRole',
+          foreign_key: :promotion_rule_id,
+          dependent: :destroy
+
+        has_many :roles,
+          through: :promotion_rule_roles,
+          class_name: 'Spree::Role'
+
+        accepts_nested_attributes_for :roles
+
+        def applicable?(promotable)
+          promotable.is_a?(Spree::Order)
+        end
+
+        def eligible?(order, options = {})
+          (roles & order.user.spree_roles).any? if order.user
```

Comment:
I think what's here is probably fine. Both solutions run the same number (2) of SQL queries and I don't expect instantiation of a few roles to be too expensive.

If we were really concerned about performance. We could store the role ids as a preference (`preferred_role_ids`) instead of as a real association. This would allow us to avoid having N+1 queries if we were to have several role-based promotions which were automatically applied.